File: app/services/auth_service.py
"""Password hashing and session management — stdlib only (hashlib/hmac/
secrets), no new dependency for what's otherwise a small amount of logic
(CLAUDE.md §10: "every new dependency is a decision, not a default").

Sessions are opaque random tokens stored server-side (a `sessions`
collection), not signed/self-contained tokens like a JWT — simpler to reason
about and revoke (delete the row) at this system's scale, and MongoDB's own
TTL index (see app/core/database.py) expires them without a scheduled job."""

import hashlib
import hmac
import logging
import secrets
from datetime import datetime, timedelta, timezone

from pymongo.errors import DuplicateKeyError

from app.core.config import settings
from app.core.database import sessions, users

logger = logging.getLogger(__name__)

_PBKDF2_ITERATIONS = 260_000


def hash_password(password: str) -> str:
    salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac("sha256", password.encode(), bytes.fromhex(salt), _PBKDF2_ITERATIONS)
    return f"{salt}${digest.hex()}"


def verify_password(password: str, stored_hash: str) -> bool:
    try:
        salt, digest_hex = stored_hash.split("$", 1)
    except ValueError:
        return False
    candidate = hashlib.pbkdf2_hmac("sha256", password.encode(), bytes.fromhex(salt), _PBKDF2_ITERATIONS)
    return hmac.compare_digest(candidate.hex(), digest_hex)


async def authenticate(username: str, password: str) -> dict | None:
    """Returns the user document on success, None on any failure — a caller
    never needs to distinguish "no such user" from "wrong password" (that
    distinction is exactly what lets an attacker enumerate usernames)."""
    user = await users.find_one({"username": username})
    if user is None:
        return None
    if user.get("status") != "active":
        return None
    if not verify_password(password, user["password_hash"]):
        return None
    return user


async def create_session(user: dict) -> str:
    token = secrets.token_urlsafe(32)
    now = datetime.now(timezone.utc)
    await sessions.insert_one(
        {
            "_id": token,
            "user_id": str(user["_id"]),
            "username": user["username"],
            "role": user["role"],
            "created_at": now,
            "expires_at": now + timedelta(hours=settings.session_ttl_hours),
        }
    )
    return token


async def get_session(token: str) -> dict | None:
    session = await sessions.find_one({"_id": token})
    if session is None:
        return None
    # Belt-and-suspenders: the TTL index reaps expired sessions in the
    # background (typically within ~60s of expiry, not instantly), so a
    # request landing in that window must not be treated as authenticated.
    if session["expires_at"] <= datetime.now(timezone.utc):
        return None
    return session


async def delete_session(token: str) -> None:
    await sessions.delete_one({"_id": token})


async def bootstrap_initial_admin() -> None:
    """Create one admin account if the users collection is completely
    empty — otherwise there'd be no way to log in and create the first
    user. If INITIAL_ADMIN_PASSWORD isn't set, a random one is generated
    and logged at WARNING level (never silently) so it's impossible to miss
    but still meant to be changed immediately via the Users page.

    gunicorn runs multiple worker processes, each running this at startup
    independently (same reasoning as get_global_settings's race fix) — on a
    genuinely fresh database, more than one worker can see zero users and
    try to create the same username at once. The count_documents check
    above is just a fast path for every startup after the first; the actual
    race safety is catching DuplicateKeyError below, so only the worker
    that actually won gets to log its password as the real one."""
    if await users.count_documents({}, limit=1):
        return

    password = settings.initial_admin_password or secrets.token_urlsafe(12)
    now = datetime.now(timezone.utc)
    try:
        await users.insert_one(
            {
                "username": settings.initial_admin_username,
                "password_hash": hash_password(password),
                "role": "admin",
                "status": "active",
                "created_at": now,
                "updated_at": now,
            }
        )
    except DuplicateKeyError:
        return  # another worker won this race and already created it

    if settings.initial_admin_password:
        logger.info("auth.bootstrap_admin_created username=%s", settings.initial_admin_username)
    else:
        # Deliberately NOT logger.warning(): the root logger's handlers
        # (see app/core/logging_config.py) include a RotatingFileHandler,
        # which would persist this plaintext password into logs/app.log
        # indefinitely — a straight violation of "credentials are never
        # logged" (CLAUDE.md §7). A one-time plain print() reaches the same
        # console an operator watches on first boot without ever touching
        # the log file.
        print(
            f"auth.bootstrap_admin_created username={settings.initial_admin_username} "
            f"password={password} — no INITIAL_ADMIN_PASSWORD was set, so this was "
            "generated. Log in and change it from the Users page now. This password "
            "will not be shown again and is not written to any log file.",
            flush=True,
        )

File: lib/services/auth_service.dart
import 'dart:convert';
import 'dart:math';

import 'package:drift/drift.dart';
import 'package:pointycastle/export.dart';

import '../core/errors.dart';
import '../core/logging.dart';
import '../core/role.dart';
import '../data/local/database.dart';

/// Password hashing and opaque server-side sessions — a port of v1
/// `app/services/auth_service.py`. PBKDF2-HMAC-SHA256, 260k iterations, stored
/// as `salt$digest` (both hex), matching v1 so a migrated `users` row verifies
/// unchanged. Sessions are random tokens in a table row, revoked by deleting
/// the row; expired rows are swept opportunistically (no cron), the same
/// reasoning as v1's Mongo TTL index (PRD §8).
class AuthService {
  AuthService(this._db, {required this.sessionTtl});

  final AppDatabase _db;
  final Duration sessionTtl;
  final _log = log('auth');

  static const _iterations = 260000;
  static const _keyLen = 32;
  final _random = Random.secure();

  /// Throttle for [sweepExpired] — the web admin fires many requests per
  /// screen and a DELETE-scan on every one is wasted work.
  DateTime? _lastSweep;
  static const _sweepEvery = Duration(minutes: 5);

  // ---- password hashing -------------------------------------------------

  String hashPassword(String password) {
    final salt = _randomBytes(16);
    final digest = _pbkdf2(password, salt);
    return '${_hex(salt)}\$${_hex(digest)}';
  }

  bool verifyPassword(String password, String storedHash) {
    final parts = storedHash.split('\$');
    if (parts.length != 2) return false;
    final salt = _unhex(parts[0]);
    final expected = parts[1];
    final actual = _hex(_pbkdf2(password, salt));
    return _constantTimeEquals(actual, expected);
  }

  Uint8List _pbkdf2(String password, Uint8List salt) {
    final derivator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64))
      ..init(Pbkdf2Parameters(salt, _iterations, _keyLen));
    return derivator.process(Uint8List.fromList(utf8.encode(password)));
  }

  // ---- authentication + sessions -------------------------------------

  /// Returns the user row on success, null on ANY failure — a caller never
  /// distinguishes "no such user" from "wrong password" (that distinction is
  /// what lets an attacker enumerate usernames). Same as v1.
  Future<User?> authenticate(String username, String password) async {
    final user = await (_db.select(_db.users)
          ..where((u) => u.username.equals(username)))
        .getSingleOrNull();
    if (user == null) return null;
    if (user.status != 'active') return null;
    if (!verifyPassword(password, user.passwordHash)) return null;
    return user;
  }

  Future<Session> createSession(User user) async {
    final now = DateTime.now().toUtc();
    final token = _urlSafeToken(32);
    final row = SessionsCompanion.insert(
      token: token,
      userId: user.id,
      username: user.username,
      role: user.role,
      createdAt: now,
      expiresAt: now.add(sessionTtl),
    );
    await _db.into(_db.sessions).insert(row);
    _log.info('login_succeeded username=${user.username} role=${user.role}');
    return (await _sessionByToken(token))!;
  }

  /// Resolves a bearer token to its session, or throws [AuthException].
  /// Belt-and-suspenders expiry check: a token past `expiresAt` is rejected
  /// even before the sweep removes its row.
  Future<Session> requireSession(String? token) async {
    if (token == null || token.isEmpty) {
      throw const AuthException('Not logged in.');
    }
    final session = await _sessionByToken(token);
    if (session == null || session.expiresAt.isBefore(DateTime.now().toUtc())) {
      throw const AuthException(
          'Session expired or invalid — please log in again.');
    }
    return session;
  }

  Future<void> deleteSession(String token) async {
    await (_db.delete(_db.sessions)..where((s) => s.token.equals(token))).go();
  }

  /// Opportunistic cleanup — safe to call from the auth middleware on every
  /// request; it self-throttles to once per [_sweepEvery] so it isn't a DB
  /// write per API call. No scheduled job.
  Future<void> sweepExpired() async {
    final now = DateTime.now().toUtc();
    if (_lastSweep != null && now.difference(_lastSweep!) < _sweepEvery) return;
    _lastSweep = now;
    await (_db.delete(_db.sessions)
          ..where((s) => s.expiresAt.isSmallerThanValue(now)))
        .go();
  }

  Future<Session?> _sessionByToken(String token) =>
      (_db.select(_db.sessions)..where((s) => s.token.equals(token)))
          .getSingleOrNull();

  // ---- first-run setup -------------------------------------------------

  /// True while this host has no accounts at all — the state the setup screen
  /// exists to resolve. Also true after a data reset, which is what makes
  /// setup self-healing rather than a one-shot that can be missed.
  Future<bool> needsSetup() async => (await _db.users.count().getSingle()) == 0;

  /// Creates the first admin account with credentials the operator chose.
  ///
  /// This replaced a scheme that generated a password and displayed it once on
  /// the login screen. The account was real, but the password lived only in
  /// memory: background the app, restart it, or simply not read the banner,
  /// and the host was left with an admin nobody could ever sign in as and no
  /// way back except wiping the database. An operator who picks their own
  /// password cannot be locked out by a missed notification.
  ///
  /// Guarded on an empty users table so it can never be used to mint a second
  /// admin on a live host.
  Future<void> createInitialAdmin({
    required String username,
    required String password,
  }) async {
    if (!await needsSetup()) {
      throw const ValidationException(
          'This device is already set up. Sign in instead.');
    }
    final name = username.trim();
    if (name.isEmpty) {
      throw const ValidationException('Choose a username.');
    }
    requireStrongPassword(password);

    final now = DateTime.now().toUtc();
    await _db.into(_db.users).insert(UsersCompanion.insert(
          username: name,
          passwordHash: hashPassword(password),
          role: Role.admin.wire,
          createdAt: now,
          updatedAt: now,
        ));
    _log.info('initial_admin_created username=$name');
  }

  /// Rejects passwords that would make the host trivially reachable by anyone
  /// on the same Wi-Fi. Deliberately a length floor rather than a character
  /// -class rule: length is what actually helps, and rules people fight
  /// produce `Password1!` (CLAUDE.md §11.2, Postel's Law on input).
  static const minPasswordLength = 8;

  void requireStrongPassword(String password) {
    if (password.length < minPasswordLength) {
      throw const ValidationException(
          'Use at least $minPasswordLength characters.');
    }
  }

  /// A strong password the operator can accept rather than invent. Returned,
  /// never logged — a plaintext credential in the rotating log file would
  /// violate CLAUDE.md §7.
  String suggestPassword() => _urlSafeToken(9);

  /// Last-resort recovery, host device only (see the route's role gate).
  ///
  /// Grants no privilege that physical possession of the host phone did not
  /// already carry: that phone can already wipe the entire database from
  /// Settings, which is strictly more destructive than a password reset.
  Future<void> resetPasswordFor(String username, String newPassword) async {
    requireStrongPassword(newPassword);
    final updated = await (_db.update(_db.users)
          ..where((u) => u.username.equals(username)))
        .write(UsersCompanion(
      passwordHash: Value(hashPassword(newPassword)),
      updatedAt: Value(DateTime.now().toUtc()),
    ));
    if (updated == 0) {
      throw ValidationException('No account called "$username".');
    }
    // Any session opened with the old password stops being valid.
    await (_db.delete(_db.sessions)..where((s) => s.username.equals(username)))
        .go();
    _log.warning('admin_password_reset username=$username');
  }

  // ---- primitives --------------------------------------------------------

  Uint8List _randomBytes(int n) =>
      Uint8List.fromList(List.generate(n, (_) => _random.nextInt(256)));

  String _urlSafeToken(int bytes) =>
      base64Url.encode(_randomBytes(bytes)).replaceAll('=', '');

  static String _hex(Uint8List b) =>
      b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

  static Uint8List _unhex(String s) {
    final out = Uint8List(s.length ~/ 2);
    for (var i = 0; i < out.length; i++) {
      out[i] = int.parse(s.substring(i * 2, i * 2 + 2), radix: 16);
    }
    return out;
  }

  static bool _constantTimeEquals(String a, String b) {
    if (a.length != b.length) return false;
    var diff = 0;
    for (var i = 0; i < a.length; i++) {
      diff |= a.codeUnitAt(i) ^ b.codeUnitAt(i);
    }
    return diff == 0;
  }
}
